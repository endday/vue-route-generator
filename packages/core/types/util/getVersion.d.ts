export default function main(): {
    isVue2: boolean;
    isVue3: boolean;
};
