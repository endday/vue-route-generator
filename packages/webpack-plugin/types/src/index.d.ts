import type { Compiler } from 'webpack';
import { GenerateConfig } from '@auto-route/core';
export interface PluginConfig extends GenerateConfig {
    pages: string;
    outFile?: string;
}
export declare class Plugin {
    options: PluginConfig;
    hasRun: boolean;
    modules: string[];
    folderSegments: string[];
    private watcher;
    constructor(options: PluginConfig);
    private initWatcher;
    apply(compiler: Compiler): void;
}
