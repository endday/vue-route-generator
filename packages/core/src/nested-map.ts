export interface NestedMap<T> {
  value?: T
  children?: Map<string, NestedMap<T>>
}

export function setToMap<T>(map: NestedMap<T>, path: string[], value: T): void {
  const target = path.reduce((item, key) => {
    if (!item.children) {
      item.children = new Map()
    }

    let child = item.children.get(key)
    if (!child) {
      child = {}
      item.children.set(key, child)
    }

    return child
  }, map)

  target.value = value
}
