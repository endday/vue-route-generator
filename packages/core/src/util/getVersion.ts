const packageJson = require('../../package.json')

export default function main() {
  let vue = null
  const data = {
    isVue2: false,
    isVue3: false
  }
  try {
    vue = require('vue')
  } catch (e) {
    console.log(`[${packageJson.name}] not current Vue version, please use Vue2/3`)
    return data
  }
  const { version } = vue

  if (typeof version !== 'string' || !(version.startsWith('2.') || version.startsWith('3.'))) {
    console.log(`[${packageJson.name}] not current Vue version, please use Vue2/3`)
    return data
  }

  data.isVue2 = version.startsWith('2.')
  data.isVue3 = version.startsWith('3.')

  return data
}
