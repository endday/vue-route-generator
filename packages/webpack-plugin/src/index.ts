import * as fs from 'fs'
import * as path from 'path'
import * as chokidar from 'chokidar'
import type { FSWatcher } from 'chokidar'
import type { Compiler } from 'webpack'
import { generateRoutes, GenerateConfig } from '@auto-route/core'

const pluginName = 'AutoRoutingPlugin'

export interface PluginConfig extends GenerateConfig {
	outFile?: string
}

export class Plugin {
	options: PluginConfig
	hasRun: boolean
	modules: string[]
	folderSegments: string[]
	private watcher: FSWatcher

	constructor(options: PluginConfig) {
		this.options = options
		this.hasRun = false
		this.modules = []
		// 监听目录的层级 ./src/views
		this.folderSegments = this.options.pages.split('/').filter(folder => /[A-z]/.test(folder))
		this.watcher = this.initWatcher()
	}

	private initWatcher() {
		if (this.watcher) {
			return this.watcher
		}
		return chokidar.watch(this.options.pages)
	}

	apply(compiler: Compiler) {
		compiler.hooks.thisCompilation.tap(pluginName, compilation => {
			this.hasRun = true
			const generate = () => {
				const code = generateRoutes(this.options)
				const to = this.options.outFile || path.resolve(__dirname, '../index.js')
				if (
					fs.existsSync(to) &&
					fs.readFileSync(to, 'utf8').trim() === code.trim()
				) {
					return
				}

				fs.writeFileSync(to, code)
			}
			this.watcher.on('all', () => {
				try {
					generate()
				} catch (error: any) {
					compilation.errors.push(error)
				}
			})
		})
	}
}
