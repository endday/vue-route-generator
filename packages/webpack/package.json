{
  "name": "@vue-auto-router/webpack",
  "version": "1.0.0",
  "main": "./lib/index.es.js",
  "types": "./types/index.d.ts",
  "files": [
    "lib"
  ],
  "exports": {
    ".": {
      "require": "./lib/index.cjs.js",
      "import": "./lib/index.es.js"
    }
  },
  "scripts": {
    "build": "vite build",
    "test": "vitest",
    "test-update": "vitest -u",
    "prebuild": "npm run test"
  },
  "dependencies": {
    "@vue-auto-router/core": "workspace:*"
  },
  "devDependencies": {
    "@types/fs-extra": "^9.0.6",
    "fs-extra": "^9.0.1",
    "webpack": "^5.11.1",
    "chokidar": "^3.5.3"
  }
}
