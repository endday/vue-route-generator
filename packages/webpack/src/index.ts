import * as fs from 'fs'
import { resolve } from 'path'
import * as chokidar from 'chokidar'
import type { FSWatcher } from 'chokidar'
import type { Compiler } from 'webpack'
import { generateRoutes, GenerateConfig } from '@vue-auto-router/core'

const pluginName = 'AutoRoutingPlugin'

export interface PluginConfig extends GenerateConfig {
  pages: string,
  outFile?: string,
  useWatcher?: boolean
}

export class Plugin {
  options: PluginConfig
  modules: string[]
  folderSegments: string[]
  private watcher: FSWatcher | undefined

  constructor(options: PluginConfig) {
    this.options = options
    this.modules = []
    // 监听目录的层级 ./src/views
    this.folderSegments = this.options.pages.split('/').filter(folder => /[A-z]/.test(folder))
    if (this.options.useWatcher) {
      this.watcher = this.initWatcher()
    } else {
      this.watcher = undefined
    }
  }

  private initWatcher() {
    if (this.watcher) {
      return this.watcher
    }
    return chokidar.watch(this.options.pages)
  }

  apply(compiler: Compiler) {
    const generate = () => {
      const code = generateRoutes(this.options)
      const to = this.options.outFile || resolve(__dirname, '../router.js')
      if (
        fs.existsSync(to) &&
        fs.readFileSync(to, 'utf8').trim() === code.trim()
      ) {
        return
      }
      fs.writeFileSync(to, code)
    }
    compiler.hooks.thisCompilation.tap(pluginName, compilation => {
      if (this.watcher && this.options.useWatcher) {
        this.watcher.on('all', () => {
          try {
            generate()
          } catch (error: any) {
            compilation.errors.push(error)
          }
        })
      } else {
        try {
          generate()
        } catch (error: any) {
          compilation.errors.push(error)
        }
      }
    })
  }
}
