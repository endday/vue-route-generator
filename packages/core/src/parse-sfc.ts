import getVersion from './util/getVersion'
import type { parseComponent as _parserV2 } from 'vue-template-compiler'
import type { parse as _parserV3 } from '@vue/compiler-sfc'

const packageJson = require('../../package.json')

const { isVue2 } = getVersion()

export interface ParseResult {
  customBlocks: CustomBlock[]
}

export interface CustomBlock {
  type: string
  content: string
}

function parserV2Fn(code: string) {
  const parserV2: typeof _parserV2 = require('vue-template-compiler')
    .parseComponent
  return parserV2(code, {
    pad: 'space'
  })
}

function parserV3Fn(code: string) {
  const parserV3: typeof _parserV3 = require('@vue/compiler-sfc').parse
  return parserV3(code, {
    pad: 'space'
  }).descriptor
}

export function parseSFC(code: string): ParseResult {
  try {
    if (isVue2) {
      return parserV2Fn(code)
    } else {
      return parserV3Fn(code)
    }
  } catch {
    throw new Error(
      `[${packageJson.name}] Either "vue-template-compiler" or "@vue/compiler-sfc" is required.`
    )
  }
}
