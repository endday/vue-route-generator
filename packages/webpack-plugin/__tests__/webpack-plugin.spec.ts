import { describe, expect, beforeEach, it } from 'vitest'
import * as path from 'path'
import * as fse from 'fs-extra'
import { webpack } from 'webpack'
import type { Compiler } from 'webpack'
import { Plugin } from '../src'
import { getExternals } from '../../../scripts/util'
import { builtinModules } from 'module'

const resolve = (p: string): string => path.resolve(__dirname, p)

const compiler = (plugin: Plugin): Compiler => {
  return webpack({
    mode: 'none',
    entry: resolve('./fixtures/fake-router.js'),
    output: {
      path: resolve('./fixtures/out'),
      filename: 'main.js'
    },
    resolve: {
      alias: {
        '@': resolve('./fixtures/')
      }
    },
    externals: [...getExternals(), ...builtinModules],
    plugins: [plugin]
  })
}

const matchOutputWithSnapshot = (path?: string) => {
  const out = fse.readFileSync(
    resolve(path || './fixtures/router.js'),
    'utf8'
  )
  expect(out).toMatchSnapshot()
}

const addPage = (p: string, content = '') => {
  const to = resolve(path.join('fixtures/pages', p))
  fse.outputFileSync(to, content)
}

const removePage = (p: string) => {
  const to = resolve(path.join('fixtures/pages', p))
  fse.removeSync(to)
}

describe('webpack plugin', () => {
  beforeEach(() => {
    fse.removeSync(resolve('../router.js'))

    // reset pages
    fse.removeSync(resolve('fixtures/pages'))
    addPage('index.vue')
    addPage('users/foo.vue')
    addPage('users/_id.vue')
  })

  it('imports dynamically created routes', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    compiler(plugin).run(() => {
      matchOutputWithSnapshot()
      done('done')
    })
  }))

  it('watches adding a page', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    let count = 0
    const watching = compiler(plugin).watch({}, () => {
      count++
      switch (count) {
        case 1:
          addPage('users.vue')
          break
        default:
          matchOutputWithSnapshot()
          watching.close(done)
      }
    })
  }))

  it('watches changing route custom block data', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    let count = 0
    const watching = compiler(plugin).watch({}, () => {
      count++
      switch (count) {
        case 1:
          addPage(
            'users/foo.vue',
            `
              <route>
              {
                "requiresAuth": true
              }
              </route>
            `
          )
          break
        default:
          matchOutputWithSnapshot()
          watching.close(done)
      }
    })
  }))

  it('watches removing a page', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    let count = 0
    const watching = compiler(plugin).watch({}, () => {
      count++
      if (count === 1) {
        removePage('users/foo.vue')
      } else {
        matchOutputWithSnapshot()
        watching.close(done)
      }
    })
  }))

  it('does not fire compilation when the route does not changed', () => new Promise((done, fail) => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    let count = 0
    const watching = compiler(plugin).watch({}, () => {
      count++
      if (count === 10) {
        fail('webpack watcher seems to go infinite loop')
      }
    })

    setTimeout(() => {
      watching.close(done)
    }, 5000)
  }), 10000)

  it('should write to custom output file if specified', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/out/custom.js')
    })

    compiler(plugin).run(() => {
      expect(fse.existsSync(resolve('./fixtures/out/custom.js'))).toBeTruthy()
      matchOutputWithSnapshot('fixtures/out/custom.js')
      done('done')
    })
  }))

  it('should not stop watching after detecting route custom block syntax errors', () => new Promise(done => {
    const plugin = new Plugin({
      pages: resolve('fixtures/pages'),
      outFile: resolve('fixtures/router.js')
    })

    let count = 0
    const watching = compiler(plugin).watch({}, () => {
      count++
      if (count === 1) {
        addPage(
          'users/foo.vue',
          `
              <route>
              {
                "meta": {
                  "requiresAuth": true,
                }
              }
              </route>
            `
        )
      } else if (count === 2) {
        addPage(
          'users/foo.vue',
          `
              <route>
              {
                "meta": {
                  "requiresAuth": true
                }

              </route>
            `
        )
      } else if (count === 3) {// Somehow, changing content triggers compilation twice.
      } else {
        matchOutputWithSnapshot()
        watching.close(done)
      }

    })
    setTimeout(() => {
      matchOutputWithSnapshot()
      watching.close(done)
    }, 8000)
  }), 10000)
})
