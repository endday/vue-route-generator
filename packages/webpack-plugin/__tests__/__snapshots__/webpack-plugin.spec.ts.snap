// Vitest Snapshot v1

exports[`webpack plugin > imports dynamically created routes 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users_foo() {
  return import(/* webpackChunkName: \\"users-foo\\" */ '@/pages/users/foo.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users-foo',
    path: '/users/foo',
    component: users_foo,
  },
  {
    name: 'users-id',
    path: '/users/:id?',
    component: users__id,
  },
]
"
`;

exports[`webpack plugin > should not stop watching after detecting route custom block syntax errors 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users_foo() {
  return import(/* webpackChunkName: \\"users-foo\\" */ '@/pages/users/foo.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users-foo',
    path: '/users/foo',
    component: users_foo,
  },
  {
    name: 'users-id',
    path: '/users/:id?',
    component: users__id,
  },
]
"
`;

exports[`webpack plugin > should write to custom output file if specified 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users_foo() {
  return import(/* webpackChunkName: \\"users-foo\\" */ '@/pages/users/foo.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users-foo',
    path: '/users/foo',
    component: users_foo,
  },
  {
    name: 'users-id',
    path: '/users/:id?',
    component: users__id,
  },
]
"
`;

exports[`webpack plugin > watches adding a page 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users() {
  return import(/* webpackChunkName: \\"users\\" */ '@/pages/users.vue')
}
function users_foo() {
  return import(/* webpackChunkName: \\"users-foo\\" */ '@/pages/users/foo.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users',
    path: '/users',
    component: users,
    children: [
      {
        name: 'users-foo',
        path: 'foo',
        component: users_foo,
      },
      {
        name: 'users-id',
        path: ':id?',
        component: users__id,
      },
    ],
  },
]
"
`;

exports[`webpack plugin > watches changing route custom block data 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users_foo() {
  return import(/* webpackChunkName: \\"users-foo\\" */ '@/pages/users/foo.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users-foo',
    path: '/users/foo',
    component: users_foo,
    requiresAuth: true,
  },
  {
    name: 'users-id',
    path: '/users/:id?',
    component: users__id,
  },
]
"
`;

exports[`webpack plugin > watches removing a page 1`] = `
"function index() {
  return import(/* webpackChunkName: \\"index\\" */ '@/pages/index.vue')
}
function users__id() {
  return import(/* webpackChunkName: \\"users-id\\" */ '@/pages/users/_id.vue')
}

export default [
  {
    name: 'index',
    path: '/',
    component: index,
  },
  {
    name: 'users-id',
    path: '/users/:id?',
    component: users__id,
  },
]
"
`;
